src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go

```diff
@@ -0,0 +1,235 @@
+func itemToHTML(feed *gofeed.Feed, item gofeed.Item) matrix.HTMLMessage {
+	return matrix.GetHTMLMessage("m.notice", fmt.Sprintf(
+		"<i>%s</i> posted a new article: %s ( %s )",
+		html.EscapeString(feed.Title), html.EscapeString(item.Title), html.EscapeString(item.Link),
+	))
+}
+
+type feedReaderService struct {
+	types.DefaultService
+	id            string
+	serviceUserID string
+	Feeds         map[string]struct { // feed_url => { }
+		PollIntervalMins         int   `json:"poll_interval_mins"`
+		NextPollTimestampSecs    int64 // Internal: When we should poll again
+		FeedUpdatedTimestampSecs int64 // Internal: The last time the feed was updated
+	} `json:"feeds"`
+	Rooms map[string][]string `json:"rooms"` // room_id => [ feed_url ]
+}
+
+func (s *feedReaderService) ServiceUserID() string { return s.serviceUserID }
+func (s *feedReaderService) ServiceID() string     { return s.id }
+func (s *feedReaderService) ServiceType() string   { return "feedreader" }
+func (s *feedReaderService) Poller() types.Poller  { return &feedPoller{} }
+
+// Register will check the liveness of each RSS feed given. If all feeds check out okay, no error is returned.
+func (s *feedReaderService) Register(oldService types.Service, client *matrix.Client) error {
+	if len(s.Feeds) == 0 {
+		// this is an error UNLESS the old service had some feeds in which case they are deleting us :(
+		var numOldFeeds int
+		oldFeedService, ok := oldService.(*feedReaderService)
+		if !ok {
+			log.WithField("service_id", oldService.ServiceID()).Error("Old service isn't a FeedReaderService")
+		} else {
+			numOldFeeds = len(oldFeedService.Feeds)
+		}
+		if numOldFeeds == 0 {
+			return errors.New("An RSS feed must be specified.")
+		}
+		return nil
+	}
+	// Make sure we can parse the feed
+	for feedURL := range s.Feeds {
+		fp := gofeed.NewParser()
+		if _, err := fp.ParseURL(feedURL); err != nil {
+			return fmt.Errorf("Failed to read URL %s: %s", feedURL, err.Error())
+		}
+	}
+	// Make sure all feeds are accounted for (appear at least once) in the room map, AND make sure there
+	// are no weird new feeds in those rooms
+	for roomID, roomFeeds := range s.Rooms {
+		for _, f := range roomFeeds {
+			if _, exists := s.Feeds[f]; !exists {
+				return fmt.Errorf("Feed URL %s in room %s does not exist in the Feeds section", f, roomID)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *feedReaderService) PostRegister(oldService types.Service) {
+	if len(s.Feeds) == 0 { // bye-bye :(
+		logger := log.WithFields(log.Fields{
+			"service_id":   s.ServiceID(),
+			"service_type": s.ServiceType(),
+		})
+		logger.Info("Deleting service (0 feeds)")
+		polling.StopPolling(s)
+		if err := database.GetServiceDB().DeleteService(s.ServiceID()); err != nil {
+			logger.WithError(err).Error("Failed to delete service")
+		}
+	}
+}
+
+func init() {
+	types.RegisterService(func(serviceID, serviceUserID, webhookEndpointURL string) types.Service {
```

Comment:
Not relevant to this particular service, but that is the function signature of the factory method.